test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,338 @@
+    return;
+  }
+
+  if (Array.isArray(expected)) {
+    expect(input).to.be.an('array');
+    for (const [index, expectedValue] of input.entries()) {
+      compareInputToSpec(input[index], expectedValue);
+    }
+    return;
+  }
+
+  if (expected && typeof expected === 'object') {
+    for (const [expectedPropName, expectedValue] of Object.entries(expected)) {
+      expect(input).to.have.property(expectedPropName);
+      compareInputToSpec(input[expectedPropName], expectedValue);
+    }
+    return;
+  }
+
+  expect(input).to.equal(expected);
+};
+
+const getTestOpDefinitions = (threadContext: ThreadContext) => ({
+  checkOut: async function (op) {
+    const connection: Connection = await util
+      .promisify(ConnectionPool.prototype.checkOut)
+      .call(threadContext.pool);
+    if (op.label != null) {
+      threadContext.connections.set(op.label, connection);
+    } else {
+      threadContext.orphans.add(connection);
+    }
+  },
+  checkIn: function (op) {
+    const connection = threadContext.connections.get(op.connection);
+    threadContext.connections.delete(op.connection);
+
+    if (!connection) {
+      throw new Error(`Attempted to release non-existient connection ${op.connection}`);
+    }
+
+    return threadContext.pool.checkIn(connection);
+  },
+  clear: function () {
+    return threadContext.pool.clear();
+  },
+  close: async function () {
+    return await util.promisify(ConnectionPool.prototype.close).call(threadContext.pool);
+  },
+  wait: async function (options) {
+    const ms = options.ms;
+    return sleep(ms);
+  },
+  start: function (options) {
+    const target = options.target;
+    const thread = threadContext.getThread(target);
+    thread.start();
+  },
+  waitForThread: async function (options): Promise<void> {
+    const name = options.name;
+    const target = options.target;
+
+    const threadObj = threadContext.threads.get(target);
+
+    if (!threadObj) {
+      throw new Error(`Attempted to run op ${name} on non-existent thread ${target}`);
+    }
+
+    await threadObj.finish();
+  },
+  waitForEvent: function (options): Promise<void> {
+    const event = options.event;
+    const count = options.count;
+    return new Promise(resolve => {
+      function run() {
+        if (threadContext.poolEvents.filter(ev => getEventType(ev) === event).length >= count) {
+          return resolve();
+        }
+
+        threadContext.poolEventsEventEmitter.once('poolEvent', run);
+      }
+      run();
+    });
+  }
+});
+
+export class ThreadContext {
+  pool: ConnectionPool;
+  threads: Map<any, Thread>;
+  connections: Map<string, Connection>;
+  orphans: Set<Connection>;
+  poolEvents = [];
+  poolEventsEventEmitter = new EventEmitter();
+  hostAddress: HostAddress;
+  supportedOperations: ReturnType<typeof getTestOpDefinitions>;
+
+  constructor(hostAddress) {
+    this.threads = new Map();
+    this.connections = new Map();
+    this.orphans = new Set();
+    this.poolEvents = [];
```

Comment:
nit: I think these 2 lines can be removed as we are defaulting them to exactly the same objects above on line 226 & 227.

Reply:
yeah let me clean this up, most of these don't need to be set in the constructor

Reply:
Thanks. I'm a LGTM after that in case it gets late here in the evening.

Reply:
updated